woocommerce/woocommerce-android
# Implement statement descriptor sanitation

Introducing `StatementDescriptor` class which performs sanitation of the original statement descriptor in `StatementDescriptor::value` getter.



Closes: #7586

## Description
`StatementDescriptor` acts as a wrapper to raw statement descriptor text and applies a set of sanitation operations to prepare it for use with Stripe API. There are the following requirements applied:
 * Text must be max 22 chars long.
 * It must not contain the following characters: <, >, ', ".
 * Similarly to [iOS implementation](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Hardware/Hardware/CardReader/StatementDescriptor.swift) illegal chars are replaced by "-" char.

## Testing
The crash was reproduced in case statement descriptor (which by default is made of a website's domain) was too long -exceeded 22 chars in length.

To test this correctly you'd need to reproduce the crash on `trunk`:
1. Connect to the store with a descriptor name length exceeding 22 chars or containing illegal chars. 
2. Observe the crash when collecting payment using card.

Next, verify that the crash is not happening on `issue-7586/statement-descriptor-improvement` branch.
